### ISSUE TITLE
eip - diffs didn't match during apply

### ISSUE THREAD

**Issue by gwiddel** [hashicorp/terraform#6904]
Error applying plan:

1 error(s) occurred:

```
* aws_eip.ip.0: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.6.15
    Resource ID: aws_eip.ip.0
    Mismatch reason: attribute mismatch: instance
    Diff One (usually from plan): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff{"instance":*terraform.ResourceAttrDiff{Old:"i-8acf8400", New:"${element(aws_instance.cluster.*.id, count.index)}", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}}, Destroy:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff(nil), Destroy:false, DestroyTainted:false}
```

This occurred when I changed the `servers` count from 1 to 2, in the following resource:

``` hcl
module "BastionLinux" {
  source    = "git::ssh://git@github.datapipe.net/Automation/tf-instance.git"
  servers   = 2
  ami     = "${lookup(var.amzn-linux-amis, var.region)}"
  subnet_id   = "${terraform_remote_state.vpc-remote.output.public_subnets}"
  keyname   = "${var.keyname}"
  root_volume_size  = "10"
  root_volume_type  = "standard"
  name_tag    = "JumpHostName"
  instance_size   = "t2.micro"
  environment_tag   = "${var.environment}"
  security_groups  = "${module.dpmgmt.id},${aws_security_group.allow_ssh.id},${aws_security_group.allow_rdp.id}"
  eip      = 1
}
```

**Comment by apparentlymart** [comment]
Hi @gwiddel. Sorry for the trouble here.

Would you be able to also share the configuration of your `aws_eip` resource where that `servers` variable is being interpolated? It would be helpful to understand how the EIP resource changed in response to the change to the module configuration.

**Comment by milest** [comment]
I had the exact same issue.

    Terraform Version: 0.6.16
    Resource ID: aws_eip.eip.0
    Mismatch reason: attribute mismatch: instance
    Diff One (usually from plan): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff{"instance":*terraform.ResourceAttrDiff{Old:"i-c6da7056", New:"${element(aws_instance.worker.*.id, count.index)}", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}}, Destroy:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff(nil), Destroy:false, DestroyTainted:false}

* aws_eip.eip.1: Failure associating EIP: InvalidAllocationID.NotFound: The allocation ID 'eipalloc-b9b5fc83' does not exist
	status code: 400, request id: e2d3812f-5ce1-4489-be99-0e17277a111a

```
module "common" {
  source = "../common"
}

provider "aws" {
  region = "${module.common.aws_region}"
}

variable "worker_count" {
  default = 2
}

resource "aws_instance" "worker" {
  ami = "${module.common.aws_ami}"
  instance_type = "${module.common.aws_worker_type}"
  key_name = "${module.common.key_name}"
  count = "${var.worker_count}"
  tags {
        Role = "Worker"
        Environment = "Production"
    }
}

resource "aws_eip" "eip" {
  count = "${var.worker_count}"
  instance = "${element(aws_instance.worker.*.id, count.index)}"
  vpc = true
}

output "ip" {
  value = "${join("  ", aws_eip.eip.*.public_ip)}"
}
```

**Comment by Gary-Armstrong** [comment]
I think this is probably where it is best to report my own variant of this.

Background: I needed to update a user_data for AWS instances and I did not want TF to redeploy my instances.  In order to convince TF to let me do this, I put into my aws_instance resource:
```
  lifecycle {
    ignore_changes = ["user_data"]
  }
```
That generally worked but the EIPs didn't care for it.  I was able to update user_data in state without blowing up anything and so right now this is happening in plan:
```
~ aws_eip.wxmix_compute_eip
    instance: "i-95a92b01" => "${element(aws_instance.compute.*.id, count.index)}"

~ aws_eip.wxmix_head_eip
    instance: "i-97a92b03" => "${element(aws_instance.head.*.id, count.index)}"
```
Alrighty.  What about apply?
```
* aws_eip.wxmix_head_eip: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.6.16
    Resource ID: aws_eip.wxmix_head_eip
    Mismatch reason: attribute mismatch: instance
    Diff One (usually from plan): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff{"instance":*terraform.ResourceAttrDiff{Old:"i-97a92b03", New:"${element(aws_instance.head.*.id, count.index)}", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}}, Destroy:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff(nil), Destroy:false, DestroyTainted:false}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.
```
Whee!  So, what is going on with my aws_eip resources?  Perhaps most relevant:
```
  lifecycle {
    prevent_destroy = true
    ignore_changes = ["aws_instance"]
  }
```
This has been in place for a while and does what I want, which is preserve (usually!) my EIPs.  Since I really should be working on other things, I wanted to post this before I start doing crazy stuff and paint myself into a corner of misery and woe.

Considering adding to the ignore_changes part of aws_eip.  Not sure what effect that will have.

**Comment by Gary-Armstrong** [comment]
More code for the code god.  Today I incremented my instance count from 1 to 2 and then:
```
Error applying plan:

2 error(s) occurred:

* aws_eip.wxmix_compute_eip.0: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.7.3
    Resource ID: aws_eip.wxmix_compute_eip.0
    Mismatch reason: attribute mismatch: instance
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"instance":*terraform.ResourceAttrDiff{Old:"i-972bada6", New:"${element(aws_instance.compute.*.id, count.index)}", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff(nil), Destroy:false, DestroyTainted:false}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.

* aws_eip.wxmix_compute_eip.1: Failure associating EIP: InvalidAllocationID.NotFound: The allocation ID 'eipalloc-29ae3716' does not exist
       	status code: 400, request id: 33b31044-2379-407a-bba1-0daa51cd6852
```

I followed up with a plan:
```
~ aws_eip.wxmix_compute_eip.1
    instance: "" => "i-5e71e16f"
```

Then the apply succeeded.

The aws_eip code:
```
resource "aws_eip" "wxmix_compute_eip" {
  count = "${var.compute_count}"
  depends_on = [ "aws_internet_gateway.main" ]
	instance = "${element(aws_instance.compute.*.id, count.index)}"
	vpc = true
  lifecycle {
    prevent_destroy = true
    ignore_changes = ["aws_instance"]
  }
}
```

**Comment by mitchellh** [comment]
Hello! This should be fixed in later versions of Terraform (0.7.11 at time of writing). 

Its hard to know 100% for a couple reasons:

  * This is on a now much-older version of Terraform. This is our fault, not yours, for not trying sooner, but we've fixed many diff mismatch issues since then.

  * We've fixed issues with exact error messages as this in those versions. 

Please try again! If the issue persists please just open a new issue. Thanks!
